Fix reduce-by-key CUDA

bcumming opened the pull request:
Fix bug in CUDA `reduce_by_key` implementation on V100 or later GPUs.

The bug was not triggered for current use cases of the algorithm in Arbor, though it will be a problem when more than one reduction is to be performed in a single kernel invocation, which is required for accumulating both current and conductance values.

Fixes  #736.

halfflat commented on arbor/backends/gpu/reduce_by_key.hpp:
```diff
@@ -96,74 +46,73 @@ float get_from_lane(float value, unsigned lane) {
 
-        auto source_value = impl::get_from_lane(contribution, source_lane);
-        if (source_lane<rhs) {
+        T source_value = __shfl_sync(run.key_mask, contribution, source_lane);
```
It probably makes little difference, but I believe you could just use `__shfl_down_sync(..., w)` rather than compute `source_lane` explicitly.

bcumming replied:
I think you are right.

halfflat commented on arbor/backends/gpu/reduce_by_key.hpp:
```diff
@@ -11,80 +11,30 @@ namespace impl{
-
-// return the power of 2 that is less than or equal to i
+// Return the largest power of 2 that is less than i
+// Caveat: i must be non-zero.
```
Just regarding the comment: this returns the largest power of 2 no greater than i, right?

bcumming replied:
Yes.
Due to other simplifications, this is no longer needed, and has been removed.

bcumming commented on arbor/backends/gpu/reduce_by_key.hpp:
```diff
@@ -11,80 +11,30 @@ namespace impl{
+// The result for i>2^31 is invalid, because it can't be stored in an
+// unsigned 32 bit integer.
+// This isn't a problem for the use case that we will use it for, because
+// it will be used for values in the range [0, threads_per_warp]
```
This function should be removed: it isn't needed.

bcumming replied:
Removed.

bcumming commented on modcc/printer/cudaprinter.cpp:
```diff
@@ -445,7 +449,7 @@ void emit_state_update_cu(std::ostream& out, Symbol* from,
         is_minus && out << "-";
         out << from->name()
-            << ", params_." << d.data_var << ", " << index_i_name(d.index_var) << ");\n";
+            << ", params_." << d.data_var << ", " << index_i_name(d.index_var) << ", lane_mask_);\n";
```
Yes, it is. Before doing the full conversion to calculating run-length outside `reduce_by_key`, more tests need to be added to the cuda printer to ensure that it is only called on `APIMethod`s that perform reductions.

I was not comfortable using the `__activemask` intrinsic to find the mask inside `reduce_by_key`, because the docs didn't convince me that there wouldn't be divergent threads, even directly after `__syncwarp`.

bcumming commented:
Righteos, the algorithm has been simplified further, in a way that addresses all of the review comments (I think).

I also made modcc only generate a thread_mask for API Methods that write indexed variables. Not strictly neccesary, because I doubt that there are any performance overheads, but it makes for cleaner generated code.
